bindings/pyroot/cppyy/CPyCppyy/src/CPPOverload.cxx

```diff
@@ -144,6 +145,41 @@ static inline PyObject* HandleReturn(
     int ll_action = 0;
     if (result) {
 
+     // If the "smart pointer" is only an observing pointer that indicated
+     // ownership by the caller (aka. "owner pointer"), we create a new object
+     // with a raw pointer as C++ proxy and Python owning the object.
+     //
+     // Note that one can also set the special `__creates__` attribute of the
+     // CPPOverload on Python side to `True` in order to flag creating
+     // functions. However, this has at least two drawbacks:
+     //  1. This flag can't be applied at the granularity of indivirual C++ overloads
+     //  2. It's only on the Python side, so if you want to flag these
+     //     functions in C++ as well as in Python you have to do some bookkeeping
+        if(CPPInstance_Check(result)) {
+            auto cppInstance = (CPPInstance*)result;
+
+         // Any owner pointer must also be registered as a smart pointer. Like
+         // this, we have the owner pointer type available as
+         // `cppInstance->GetSmartIsA()` and the raw type as
+         // `cppInstance->ObjectIsA()`, and we don't need to do potentially
+         // expensive inspection operations to figure out the raw type from the
+         // owner pointer.
+            if (cppInstance->IsSmart()) {
+                if(Cppyy::IsOwnerPtr(cppInstance->GetSmartIsA())) {
+                 // Here we create a new Python object with the same flags that
```

Comment:
I was wondering why you need to create this Python object yourself and configure that Python owns the C++ object.
With your new mechanism, if instead of returning a `RooDataSet*` (whose Python proxy would not own) now you return a `RooFit::Owner<RooDataSet*>`, doesn't cppyy already return a Python proxy that owns what it has in its belly and therefore would delete the C++ object when no more references are left for the Python proxy?

Reply:
Hi @etejedor, thanks for your comment! I don't quite understand your question, I'm now so fluent with the cppyy terminology...

If I understand correctly we have separate objects:
1. The Python Proxy of type `PyObject *`
2. The C++ proxy that is owned by the Python proxy which points to the actual object, can be of type raw pointer `RooDataSet *` or any smart pointer known to cppyy (e.g. `unique_ptr<RooDataSet>` or now `RooFit::Owner<RooDataSet>`
3. The actual C++ object owned by the C++ proxy, or by Python if `kIsOwner` is true

What I want to achieve here is to
3) Tell the `PyObject` that is owns the C++ object, not only the C++ proxy
2) Replace the `RooFit::Owner<RooDataSet>` proxy with a raw pointer proxy `RooDataSet *` such that there is no confusing `owned by RooFit::Owner<RooDataSet>` in the message when you `print` the Python object. But this is not the main point, only cosmetics.

If I understand your question...
> doesn't cppyy already return a Python proxy that owns what it has in its belly and therefore would delete the C++ object when no more references are left for the Python proxy?
...correctly, you are wondering why the C++ object doesn't get deleted when I delete the original PyObject, right? Well, only the C++ proxy object (in this case the `RooFit::Owner<RooDataSet*>`) gets deleted. But the `RooFit::Owner` doesn't actually own, it only indicates ownership by the caller. That's why the C++ object survives.